Improve update checking

DanTup opened the pull request:
This change enables pinging the server to check for updates regardless of whether the local version is "out of date". The server code already has a 7-day cache so the result is that we can now ping the server once every 7 days instead of waiting for the local install to be 4 weeks out of date before pinging.

The original 4 week period is still used for when we'll start warning the user they're out of date if we could not confirm with the server whether there's a new version.

Please see the annotations on the PR code below for things I'm unsure about or would like specific feedback on. Thanks!

Fixes #14920.

DanTup committed:
Improve update checking

This change emables pinging the server to check for updates regardless of whether the local version is "out of date". The server code already has a 7-day cache so the result is that we can now ping the server once every 7 days instead of waiting for the local install to be 4 weeks out of date before pinging.

The original 4 week period is still used for when we'll start warning the user they're out of date if we could not confirm with the server whether there's a new version.

DanTup commented on packages/flutter_tools/lib/src/version.dart:
```diff
@@ -236,20 +236,33 @@ class FlutterVersion {
 
+    // Get whether there's a newer version on the remote. This only goes
+    // to the server if we haven't checked recently so won't happen often.
+    // Note: This must come before the stamp load below because it also reads/writes
```
While working on the tests, I noticed that newerFrameworkVersionAvailable used to be called after we loaded the stamp, and then later on the stamp is saved. This meant the changes made in `newerFrameworkVersionAvailable` would be overwritten by this stale version in memory.

It's possible that this didn't matter before, but it does not (one of the tests does two checks, and the last server ping timestamp didn't persist between checks) so I had to move it up here. Another option would be passing `stamp` into `newerFrameworkVersionAvailable`. This way seemed simplest, but not sure I like that re-ordering these two lines breaks a test.

tvolkert replied:
Yeah let's pass the stamp to the function.

DanTup replied:
Actually, this turns out to not quite fix it. Inside `_getLatestAvailableFlutterVersion` we have:

```dart
await versionCheckStamp.store(
  newTimeVersionWasChecked: _clock.now(),
  newKnownRemoteVersion: remoteFrameworkCommitDate,
);
```

So the version that was loaded outside is always going to be stale if we load it before we call this method. Feels a little wonky, but not sure how to fix in a way that doesn't involve threading the same stamp instance all the way through (including in the test calls). WDYT?

tvolkert replied:
Hmm.  Yeah, sounds like threading it all the way through is the way to go.

DanTup commented on packages/flutter_tools/lib/src/version.dart:
```diff
@@ -236,20 +236,33 @@ class FlutterVersion {
-    if (beenAWhileSinceWarningWasPrinted && installationSeemsOutdated && await newerFrameworkVersionAvailable()) {
+    // We show a warning if either we know there is a new remote version, or we couldn't tell but the local
+    // version is outdated.
+    final bool canShowWarning = isNewRemoteVersion || isNewRemoteVersion == null && installationSeemsOutdated;
```
I'm also not totally sold on using true/false/null here. I did consider an enum, but I'm not sure Unknown/UpdateAvailable/NoUpdateAvailable is all that more descriptive. WDYT?

DanTup commented on packages/flutter_tools/test/version_test.dart:
```diff
@@ -66,7 +66,14 @@ void main() {
+        mockCache,
+        localCommitDate: _upToDateVersion,
+        // Server will be pinged because we haven't pinged within last x days
+        expectServerPing: true,
```
This test never used to cause a server ping, but since it doesn't provide a local stamp, we do now ping the server so I had to tweak the expectations. I could also have provided a local stamp, but it felt like that was covered better in the other tests already.

DanTup commented on packages/flutter_tools/test/version_test.dart:
```diff
@@ -201,13 +208,13 @@ void main() {
     });
 
-    testUsingContext('ignores network issues', () async {
+    testUsingContext('does not print warning when unable to connect to server if not out of date', () async {
```
There are now two possibilities when there are network issues - if the local repo is stale we'll still warn, but if it's not, we will not. So I duplicated this test to handle both; the first one has the same expectation (no warning) but uses a fresh local version, and the second one has an out of date repo and therefore expects the warning.

DanTup committed:
Improve message when we know there's a new version available

DanTup commented:
I've also pushed an update that tweaks the text to say there's an update available when we know that's true. It still uses the original ("You're x days old") text when the reason for the message is that we're out of date but couldn't contact the server (shown only after 4 weeks currently).

DanTup committed:
Fix bnullable bool checks

DanTup commented:
This happened on the CI:

```
Failed assertion: boolean expression must not be null
#0      FlutterVersion.checkFlutterVersionFreshness (package:flutter_tools/src/version.dart:263:33)
```

I presume it's because I have `if (nullableBool)` instead of `if (nullableBool == true)`, however, I don't see any warnings or errors locally. Is there something I can do to see these same issues as the CI locally?

tvolkert commented:
> Is there something I can do to see these same issues as the CI locally?

It's probably due to differences in your repo setup.  Maybe try creating a separate clone of the repo without referencing your fork, and then patch in this diff locally?

Also, in cases where you have a nullable boolean, if you can't ensure that it's not nullable (which is the preferred path), then you can also say `if (nullableBoolean ?? false) {`

tvolkert commented on packages/flutter_tools/lib/src/version.dart:
```diff
@@ -236,21 +236,37 @@ class FlutterVersion {
 
+    /// Gets whether or not there is a new version of Flutter available on the remote.
+    ///
+    /// Returns null if the cached version is out-of-date or missing, and we are
```
If this can return tri-state, let's use an enum rather than a nullable boolean.

DanTup replied:
Done; reads much nicer now 👍

DanTup commented:
> It's probably due to differences in your repo setup.

It was basically a hard reset on current master when I did it. I wondered if maybe the scripts were passing some stricter flags or something that I'm not getting through normal analysis. I'll do some digging - I think this is the second time I've had no warnings locally and then seen an error like this on CI...

> `if (nullableBoolean ?? false)`

Ah, good call. I'd done `== true`, but I prefer this way because it seems less likely someone will come and delete the `== true` thinking it's not doing anything. I'll tweak this :-)

DanTup committed:
Switch nullable bool to enum

DanTup committed:
Fix casing of enum values

DanTup committed:
Remove stale comment

The names are now descriptive so doesn't need additional explanation.

DanTup commented:
@tvolkert I've fixed the minor issues, but not sure how best to handle this:

> Yeah, sounds like threading it all the way through is the way to go.

The problem is if we reverse the other of those lines with the comment, we end up with this:

- Load the stamp
- Call a method that checks the server, which needs to update the stamp

Even if we pass the stamp in to this method, it's immutable so the method would need to store it, load a new copy, then pass that one back out:

```
stamp = _getLatestAvailableFlutterVersion(stamp);
```

This feels like a bit of a burden (and if you forget to assign the return value, it could be a hard to spot bug).

WDYT? Some possible thoughts:

- Leave it as-is, and it's a rule that you should avoid holding stamps (on the basis that they're immutable and become stale if you call a method that might read/write)
- Make it mutable, so you can hold a reference to it, pass it down, and know that it's up-to-date if any method you called modified it
 - Pass it in and return it from any method that wants it - if the method doesn't modify it, it can return the same instance, otherwise it has to reload it from disk after storing and return that

DanTup commented:
@tvolkert Any thoughts on the comment above?

Hixie commented on packages/flutter_tools/lib/src/version.dart:
```diff
@@ -275,6 +292,17 @@ class FlutterVersion {
+  ║                                                                            ║
+  ║ To update to the latest version, run "flutter upgrade".                    ║
+  ╚════════════════════════════════════════════════════════════════════════════╝
+''';
```
it'd be great if we could say what the current version was, what the new version was, and what channel the user was on.

DanTup replied:
I had a go at this, but it turned into more work than I expected - we now carry around and cache both a remote commit date and also the version. I've pushed it to a branch for now so you can have a quick scan in case it's going in the wrong direction or you'd just rather it another PR for review.

https://github.com/DanTup/flutter/commit/d474a5499e6ae8a6878f92f203df6e11ab61a1d3

The tests are all passing, but there's a lot here and it's slightly complicated to manually test (without rigging code or resetting the git pointer to an old revision and just running with the changes in working dir) so it's a little risky (which I guess goes for this PR too!).

Hixie replied:
doing it in another PR (or not at all for now, if it's not tractable) is fine.

Hixie commented:
@DanTup: @tvolkert isn't available right now, but I'm happy to take over the review. Can you give me more context for the question above?

DanTup commented:
@Hixie Sure!

This PR is basically to change the update checking. Previously it was like:

- If the local repo is stale (over 30 days old?)
  AND we haven't checked the remote for a while (7 days?)
  AND the remote for this branch has a newer timestamp than we have locally
    Show "Flutter is old" message

This change makes it always go and check the remote (if it doesn't have the < 7 day cached timestamp) so that we'll tell users about updates after < 7 days instead of only 30.

I think Todd was basically happy with the OR; the outstanding comments where about this comment in the code:

> // Note: This must come before the stamp load below because it also reads/writes

We basically have some code that loads the timestamp file and then calls code that checks the server - however this also loads the timestamp, potentially modifies it, and then writes it back. Since the loaded timestamp is immutable the calling code now has a stable timestamp. We were discussing refactoring this to maybe only have one timestamp loaded and pass it around; however because it's immutable that means we have to start passing it back (as well as passing it down). IMO this might make the code more complicated for little benefit - since the timestamp is immutable it might be expected that it could get stale and you should try to read is as close to using it as possible (and not keep copies for long periods expecting them to still be valid).

WDYT?

Hixie commented:
Thanks. Several things;

- If we always hit the server, how does that work with flaky connections? Does that mean that every time the user does anything, there's a chance it'll hang waiting for the network or crash because of a proxy problem or whatever?

- If we always hit the server, what is the cached timestamp?

Maybe a different way to approach this would be for me to ask: can you describe the precise algorithm you want to follow, ignoring what we do today?

DanTup commented:
> If we always hit the server, what is the cached timestamp?

We don't always hit the server, we still cache the results for 7 days as previous (there's an existing test `does not ping server when version stamp is up-to-date`). The difference is that we no longer wait until the local commit is 30 days old before we'll ever go to the server. So previously, we'd do no update checks for 30 days, then we'd check once every 7 days. With this change we will just check once every 7 days.

> how does that work with flaky connections? Does that mean that every time the user does anything, there's a chance it'll hang waiting for the network or crash because of a proxy problem or whatever?

Hmm, this is a good question. There was already code handling this, however looking now I think it has an issue - if we fail to hit the network we may try again on every operation until it succeeds; so possibly it's going to be slow on every operation if we cannot connect at all. Possibly it already does this, but not until 30 days old.

Maybe we should write the timestamp when we fail to check too to avoid this. Maybe we should also make the whole check async and/or put short timeouts?

> can you describe the precise algorithm you want to follow, ignoring what we do today?

Sure, the intention is to be something like this:

```
if we haven't fetched remote commit for at least 7 days
    fetch remote commit date

if failed to get remote commit date
    if local version is > 30 days old
        warn user their version is old
    else
        return

if remote commit is newer than local
    show user "update available" 
```

Hopefully this matches what's going on in `checkFlutterVersionFreshness`.

DanTup commented:
As far as I can tell, the Travis failure isn't related to this - the last few master builds have been flaky - seems like some calls to `pub run test` are giving unexpected errors codes but I can't see any obvious reason why. I'll rebase on master when it seems like that's fixed.

DanTup commented:
I did some testing with no connection, and the `git fetch` call fails pretty fast like this:

> fatal: unable to access 'https://github.com/flutter/flutter.git/': Could not resolve host: github.com

I don't know how many of the possible failures will be as quick though. Maybe we should reduce the 7-day check to something lower (3-4 days?) and then always honour it even if we didn't get a successful connection?

Hixie commented on packages/flutter_tools/lib/src/version.dart:
```diff
@@ -236,21 +236,38 @@ class FlutterVersion {
 
-    Future<bool> newerFrameworkVersionAvailable() async {
+    /// Gets whether or not there is a new version of Flutter available on the remote.
+    Future<VersionCheckResult> newerFrameworkVersionAvailable() async {
```
`isNewerFrameworkVersionAvailable` might be a clearer name for this

Hixie commented on packages/flutter_tools/lib/src/version.dart:
```diff
@@ -236,21 +236,38 @@ class FlutterVersion {
 
+    // Get whether there's a newer version on the remote. This only goes
+    // to the server if we haven't checked recently so won't happen often.
+    // Note: This must come before the stamp load below because it also reads/writes
```
please remove the "Note:" prefix, it adds nothing

Hixie replied:
(see style guide which mentions this iirc)

Hixie commented on packages/flutter_tools/lib/src/version.dart:
```diff
@@ -236,21 +236,38 @@ class FlutterVersion {
     final Duration frameworkAge = _clock.now().difference(localFrameworkCommitDate);
```
is `kPauseToLetUserReadTheMessage` intentionally mutable? if so, we should rename it to not start with `k`.

DanTup replied:
I suspect it was originally immutable, though seems like it's modified in a test setup now (presumably to speed the test up). I'll remove the `k` for now, but lmk if you'd prefer another fix.

Hixie commented on packages/flutter_tools/lib/src/version.dart:
```diff
@@ -236,21 +236,38 @@ class FlutterVersion {
     final Duration frameworkAge = _clock.now().difference(localFrameworkCommitDate);
```
Let's make kVersionAgeConsideredUpToDate be 5 weeks instead of 4, since we release betas every 4 weeks. No need to push people to update too aggressively.

Hixie commented on packages/flutter_tools/lib/src/version.dart:
```diff
@@ -236,21 +236,38 @@ class FlutterVersion {
+    // to the server if we haven't checked recently so won't happen often.
+    // Note: This must come before the stamp load below because it also reads/writes
+    // the stamp and otherwise one will be stale.
+    final VersionCheckResult remoteVersion = await newerFrameworkVersionAvailable();
```
it's kind of weird to have this be done in a nested function instead of just inline.

Hixie commented on packages/flutter_tools/lib/src/version.dart:
```diff
@@ -236,21 +236,38 @@ class FlutterVersion {
+    // version is outdated.
+    final bool canShowWarning =
+      remoteVersion == VersionCheckResult.newVersionAvailable
+      || remoteVersion == VersionCheckResult.unknown && installationSeemsOutdated;
```
indenting here is weird

Hixie commented:
> I don't know how many of the possible failures will be as quick though. Maybe we should reduce the 7-day check to something lower (3-4 days?) and then always honour it even if we didn't get a successful connection?

SGTM.

> Hopefully this matches what's going on in checkFlutterVersionFreshness.

I think it might be worth refactoring the code a little to more closely match the exact structure of the pseudo-code that you pasted above, because that would make it easier to follow.

DanTup committed:
Improve name of function

DanTup committed:
Remove note:

DanTup committed:
Rename kPauseToLetUserReadTheMessage -> timeToPauseToLetUserReadTheMessage

DanTup committed:
Change kVersionAgeConsideredUpToDate to 5 weeks from 4

DanTup committed:
Inline the isNewerFrameworkVersionAvailable check

DanTup committed:
Fix indenting (?)

DanTup committed:
Fix more indenting

DanTup committed:
Rename function to be clearer it's getting the commit date

DanTup committed:
Formating tweaks

DanTup committed:
Update stamp when connection failed, and reduce time before we'll try again

Previously we would hit the server on every command if we thought we might be out of date and we never successfully connected (eg. if you're offline). This makes the stamp update even when there's a conneciton failure so that this won't happen, but reduces the time till we check again from 7 days to 3 days to compensate a little in case it was a one-off.

https://github.com/flutter/flutter/pull/18193#issuecomment-399222269

DanTup commented:
@Hixie I've pushed changes for the comments above, including updating the timestamp on failures and reducing the server ping from 7 days to 3 days. The `checkFlutterVersionFreshness` function is still a full screen of code, so not quite as readable as the pseudo code, but I think it reads much better. I wasn't sure if to extract into more functions since they'd pollute the class a little, but if you think that's preferable let me know and I'll tweak.

Btw, it occurred to me that this more-frequent check will probably result in people seeing more update notifications when working on branches than they used to (we never used to alert them until they're something like 30 days out of date, whereas now we could pick up an update within a few days); I'm not sure if this end up being noisy and we might wish to tweak this (for ex., if you're not on an official channel branch, increase the time to check). Maybe it'll be easier to figure that out once it's in master and we've used it for real for a little while though.

DanTup commented:
(the appveyor failure is this flakey test: https://github.com/flutter/flutter/issues/18737 and unrelated to my changes)

Hixie commented:
We should just probably not check at all if you're not on an official branch. We have an array of official branches in the code somewhere already.

Hixie commented on packages/flutter_tools/lib/src/version.dart:
```diff
@@ -236,26 +238,40 @@ class FlutterVersion {
+    final VersionCheckResult remoteVersionStatus =
+    // to the server if we haven't checked recently so won't happen on every
+    // command.
+        latestFlutterCommitDate == null
```
your comment got split

DanTup committed:
Fix comment

DanTup committed:
Don't perform update checks if not on an official channel
